### PR TITLE
WIP: Integration tests

### PR DESCRIPTION
**Pull request by jacobthemyth**
This set of integration tests is meant to test the boundaries of the library. In
other words, it should call the ember-data methods and assert:
1. the correct requests are sent
2. the adapter responds correctly to the responses.

Some of these tests repeat other tests not under the "unit" category, but
1. It was hard for me to reason about what had already been covered, so I
   thought starting from scratch would be easier.
2. There wasn't currently any use of mocking or fixtures that I could find,
   which made the end-to-end tests rely on Parse being up, etc.
3. Testing the full depth of the library isn't really a unit test, but an
   integration test.

I wanted to do this in the open, so this won't be ready to merge for a little
while.

### CODE REVIEW

File: tests/helpers/mock-parse.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export default function mockParse() {
```
Comment (mixonic) [inline tests/helpers/mock-parse.js]: FWIW, we usually do this kind of global mock in a single file. For example:
- https://github.com/aptible/ember-cli-aptible-shared/blob/master/test-support/helpers/mock-stripe.js#L3

And usage for tests:
- https://github.com/aptible/ember-cli-aptible-shared/blob/master/tests/test-helper.js#L20

**Comment by mixonic** [comment]
I strongly suggest using [Pretender](https://github.com/rwjblue/ember-cli-pretender) instead of mockjax. Mockjax has a few major failings, the biggest of which are that it does not reliably use a new frame of JS for responses (so sometimes they are sync) and the second that it return the same instance of the returned object over and over, so if a single test modifies the payload a following test may receive the modified payload.

**Comment by jacobthemyth** [comment]
@QuantumInformation I ran `ember init`. This doesn't include the latest version of ember or ember-data since that would introduce breaking changes. This just includes #87, i.e. ember-cli 0.2.7

**Comment by jacobthemyth** [comment]
@mixonic good point on Pretender. I played around with both sinon and mockjax, but pretender seems to be the best path forward. Thanks.

**Comment by jacobthemyth** [comment]
#84465c5 introduces Pretender and a basic test. A couples notes/questions:
1. Is there a better way than https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L13 to access the store in an integration test? I've only ever accessed the store in model tests.
2. Any thoughts on date testing for https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L29? I want to be thorough (i.e. test that if Parse returns an ISO string, the adapter converts it to a date object), but I don't like hard coding a string literal for a date in the test. I've used Timecop.js before for this sort of thing, but not sure if anybody has any better ideas.
3. Regarding https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L11 and https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/helpers/mock-parse.js#L9-L16, I couldn't figure out a better way to handle the following circumstances:
   - Pretender doesn't support external URLs
   - The unit tests require the adapter to have the full Parse URL
     Any thoughts?
4. Regarding https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/helpers/mock-parse.js#L55-L66, I know it's a kludge, but I wanted to focus on writing integration tests rather than worry about writing a pull request to Pretender. I'll circle back to it when I have time.

**Comment by mixonic** [comment]
> 1. Is there a better way than https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L13 to access the store in an integration test? I've only ever accessed the store in model tests.

This seems funky. Integration tests do not normally start an application.

I suggest something like this (so long as you are using the latest ember-qunit and ember testing helpers dependencies):

``` js
import DS from "ember-data";

module('Integration | creating objects', {
  integration: true,
  beforeEach() {
    // this line may not be needed any more
    DS._setupContainer(this.container);
    this.store = this.container.lookup('store:main');
  },

  afterEach() {
    Ember.run(this.store, 'destroy');
  }
});
```

Of course with the mock parse stuff.

> 1. Any thoughts on date testing for https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L29? I want to be thorough (i.e. test that if Parse returns an ISO string, the adapter converts it to a date object), but I don't like hard coding a string literal for a date in the test. I've used Timecop.js before for this sort of thing, but not sure if anybody has any better ideas.

Derp, no ideas about freezing time. Stub out `Date.now`?

> 1. Regarding https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/integration/creating-objects-test.js#L11 and https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/helpers/mock-parse.js#L9-L16, I couldn't figure out a better way to handle the following circumstances:
>    - Pretender doesn't support external URLs
>    - The unit tests require the adapter to have the full Parse URL
>      Any thoughts?

The adapter should allow you to override the default host with a config setting. Then in test mode we can specify the host of `""` from the `config/environment.js` file.

> 1. Regarding https://github.com/clintjhill/ember-parse-adapter/blob/84465c59aaaad2b2d39ebd2134e2fa167c17ed00/tests/helpers/mock-parse.js#L55-L66, I know it's a kludge, but I wanted to focus on writing integration tests rather than worry about writing a pull request to Pretender. I'll circle back to it when I have time.

We usually wrap pretender in a wrapper, usually [fakeServer](https://github.com/201-created/ember-cli-fake-server), for reasons just like this. I encourage you to just create a fake server that wraps pretender and use that instead of feeling like you need to monkey patch pretender. The project is very focused, and I'm not sure patches upstream that add functionality will be welcomed.

**Pull request closed without merging**